src/Microsoft.AspNetCore.SystemWebAdapters.CoreServices/HttpRequestAdapterFeature.cs

```diff
@@ -0,0 +1,191 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Web;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.WebUtilities;
+
+namespace Microsoft.AspNetCore.SystemWebAdapters;
+
+internal class HttpRequestAdapterFeature : IHttpRequestAdapterFeature, IHttpRequestFeature, IDisposable
+{
+    private readonly int _bufferThreshold;
+    private readonly long? _bufferLimit;
+    private readonly IHttpRequestFeature _other;
+
+    private Stream? _bufferedStream;
+
+    public HttpRequestAdapterFeature(IHttpRequestFeature other, int bufferThreshold, long? bufferLimit)
+    {
+        _bufferThreshold = bufferThreshold;
+        _bufferLimit = bufferLimit;
+        _other = other;
+    }
+
+    public ReadEntityBodyMode Mode { get; private set; }
+
+    public Stream GetBufferedInputStream()
+    {
+        if (Mode is ReadEntityBodyMode.Buffered or ReadEntityBodyMode.None)
+        {
+            Mode = ReadEntityBodyMode.Buffered;
+
+            _bufferedStream ??= new FileBufferingReadStream(_other.Body, _bufferThreshold, _bufferLimit, AspNetCoreTempDirectory.TempDirectoryFactory);
+
+            return _bufferedStream;
+        }
+
+        throw new InvalidOperationException("GetBufferlessInputStream cannot be called after other stream access");
+    }
+
+    Stream IHttpRequestAdapterFeature.GetBufferlessInputStream()
+    {
+        if (Mode is ReadEntityBodyMode.Bufferless or ReadEntityBodyMode.None)
+        {
+            Mode = ReadEntityBodyMode.Bufferless;
+            return GetBody();
+        }
+
+        throw new InvalidOperationException("GetBufferlessInputStream cannot be called after other stream access");
+    }
+
+    Stream IHttpRequestAdapterFeature.InputStream
+    {
+        get
+        {
+            if (Mode is ReadEntityBodyMode.Classic && _bufferedStream is not null)
+            {
+                return _bufferedStream;
+            }
+
+            throw new InvalidOperationException("InputStream must be prebuffered");
```

Comment:
We could remove this requirement and provide an API similar to what is described in https://github.com/dotnet/systemweb-adapters/issues/164 for `HttpRequest.GetInputStreamAsync()` that would do the awaiting there. This then could block and wait with the recommendation people move to the async version

Reply:
For reference, this blocks on framework so is not a real change. Non-blocking APIs are the now available `GetBufferedStream` and `GetBufferlessStream`